### ISSUE TITLE
Investigate and ensure deployment fails when Django migration fails.

### ISSUE THREAD

**Issue by mjtravers** [fecgov/fecfile-web-api#632]
The website can go down if a django migration fails during the deployment in cloud.gov. The DEV server recently went down when a Django migration failed because of missing values in some database records.

*   [x] Update the bin/run-api.sh script so that it returns a non-zero value from the bash script should the migration fail.
*   [x] Remove the "> migrate.out" part of the script so logging goes to CirlceCI and not the file system.
*   [x] Remove the "python wait_for_db.py" line if no longer necessary. (Seems odd it is called after migration)
*   [x] Test against cloud.gov environment to ensure deployment stops on bad migration

### QA Notes ###
 
  
  
  
 null

### DEV Notes ###
 
  
  
  
 null

### Design ###
 
  
  
  
 null

**Comment by dheitzer** [comment]
I was able to replicate this issue by deploying to dev a branch with a failing migration in it.  While CircleCI does not log the migration output, the failures can be seen in the cloud.gov api log as the api is repeatedly trying to restart (removing > migrate.out did not cause the logs to show up in CircleCI).  It seems that CircleCI is incorrectly listing the api as running when it's in this 'bouncing' state:

start command:   bin/run-api.sh
     state     since                  cpu    memory    disk      details
#0   running   2024-05-01T15:02:45Z   0.0%   0 of 2G   0 of 1G   
#1   running   2024-05-01T15:02:44Z   0.0%   0 of 2G   0 of 1G  

Notice that memory and disk usage are zero even though the state is running.

I attempted various things to try to get CircleCI to recognize the failure and mark the deploy as failed:

1)  modified the run-api.sh to add an exec to the last command [as suggested here ](https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#start-commands).  This is supposed to make it so the process signals are propogated up. `exec gunicorn --bind 0.0.0.0:8081 fecfiler.wsgi -w 9`
2) Tried replacing the call to run-api.sh with the body of the script and with the added exec.
3) Modified tasks.py to add `print(f"cf push output:\n {new_deploy.stdout}")` to try and get the migrations output to display in CircleCI
4) Added `health-check-type: process` to api manifest to try and get Cloud Foundry to recognize that the api is in a failed state (it defaults to port)

None of these approaches helped and CircleCI continued to list api as deployed even when the migration failed.

Next step may be to reach out to cloud.gov to see if they have had anyone with vanilla python_buildpack web applications in incorrect states on startup failure.

**Comment by mjtravers** [comment]
David did a significant amount of work on this ticket and it doesn't really lend itself to be picked up by another dev without unintentionally going down the research paths that David has already investigated. Putting ticket on hold until he returns from vacation.

**Comment by dheitzer** [comment]
Notes:

`Update the bin/run-api.sh script so that it returns a non-zero value from the bash script should the migration fail.`
It appears this is unnecessary now as the migrate command is now failing and returning a non-zero status code which causes the script itself to also exit with an error code.

**Comment by dheitzer** [comment]
regarding how to handle the scenario where a migration succeeds but one fails: As stated in the [limitations of rolling deploys documentation](https://docs.cloudfoundry.org/devguide/deploy-apps/rolling-deploy.html#limitations), this scenario may cause downtime. and may require manual intervention to rollback the database migration by [reversing migrations as necessary](https://docs.djangoproject.com/en/3.2/topics/migrations/#reversing-migrations).

**Comment by dheitzer** [comment]
I was able to reproduce this issue in dev several weeks ago, but I am no longer able to do so.  A failing migration causes the api not to start and thus the health check and deployment to fail and be reverted with zero downtime now as expected.

One thing to note is that Cloud.gov has recently updated its Cloud Foundry instances to later versions which may explain the discrepancy.  

For this ticket, I added the [Cloud Foundry recommendation ](https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#-command) of prefixing our final command with exec to ensure signals are propagated back to the app.
